File: sdk/python/src/covenant/loader.py
from __future__ import annotations

import importlib.resources
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml
from jsonschema import Draft7Validator
from pydantic import ValidationError

from covenant.models import CovenantSpec

_schema: dict[str, Any] | None = None


@dataclass
class SpecLoadError(Exception):
    """Raised when a .covenant.yaml fails to load.

    Args:
        phase: Which loading phase failed (file/yaml/schema/model).
        path: Path to the spec file.
        issues: List of error messages for this phase.
    """

    phase: str
    path: Path
    issues: list[str]

    def __str__(self) -> str:
        return f"[{self.phase.upper()}] {'; '.join(self.issues)}"


def _get_schema() -> dict[str, Any]:
    """Load the bundled covenant.schema.json once and cache it."""
    global _schema
    if _schema is None:
        data = (
            importlib.resources.files("covenant.data")
            .joinpath("covenant.schema.json")
        )
        _schema = json.loads(data.read_text(encoding="utf-8"))
    return _schema


def load_spec(path: Path) -> CovenantSpec:
    """Load and validate a .covenant.yaml through four sequential phases.

    Phase 1 -- FILE:   path exists and is readable
    Phase 2 -- YAML:   valid YAML, top-level mapping
    Phase 3 -- SCHEMA: Draft7 JSON Schema validation (all errors collected)
    Phase 4 -- MODEL:  Pydantic model_validate (skipped if Phase 3 has errors)

    Args:
        path: Path to the .covenant.yaml file.

    Returns:
        A fully-validated CovenantSpec instance.

    Raises:
        SpecLoadError: On failure at any phase, with all issues for that phase.
    """
    # Phase 1 -- FILE
    if not path.exists():
        raise SpecLoadError(phase="file", path=path, issues=[f"File not found: {path}"])
    if not path.is_file():
        raise SpecLoadError(phase="file", path=path, issues=[f"Not a file: {path}"])

    text = path.read_text(encoding="utf-8")

    # Phase 2 -- YAML
    try:
        raw: Any = yaml.safe_load(text)
    except yaml.YAMLError as e:
        raise SpecLoadError(phase="yaml", path=path, issues=[str(e)]) from e

    if not isinstance(raw, dict):
        raise SpecLoadError(
            phase="yaml",
            path=path,
            issues=["YAML must be a mapping at the top level"],
        )

    # Phase 3 -- SCHEMA
    schema = _get_schema()
    validator = Draft7Validator(schema)
    errors = sorted(validator.iter_errors(raw), key=lambda e: list(e.path))
    if errors:
        issues = [
            f"{'.'.join(str(p) for p in e.path) or '<root>'}: {e.message}"
            for e in errors
        ]
        raise SpecLoadError(phase="schema", path=path, issues=issues)

    # Phase 4 -- MODEL
    try:
        return CovenantSpec.model_validate(raw)
    except ValidationError as e:
        issues = [f"{err['loc']}: {err['msg']}" for err in e.errors()]
        raise SpecLoadError(phase="model", path=path, issues=issues) from e

File: sdk/typescript/src/loader.ts
import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import yaml from "js-yaml";
import Ajv from "ajv";
import addFormats from "ajv-formats";
import { ZodError } from "zod";
import { CovenantSpec } from "./spec.js";

// ── JSON Schema validator (cached) ────────────────────────────────────────────

let _ajv: Ajv | null = null;
let _validateFn: ReturnType<Ajv["compile"]> | null = null;

function getValidator(): ReturnType<Ajv["compile"]> {
  if (_validateFn) return _validateFn;

  const ajv = new Ajv({ allErrors: true, strict: false });
  addFormats(ajv);
  _ajv = ajv;

  const schemaPath = resolve(
    dirname(fileURLToPath(import.meta.url)),
    "../data/covenant.schema.json"
  );
  const schema = JSON.parse(readFileSync(schemaPath, "utf-8")) as object;
  _validateFn = ajv.compile(schema);
  return _validateFn;
}

// ── Error type ────────────────────────────────────────────────────────────────

export class SpecLoadError extends Error {
  constructor(
    readonly phase: "file" | "yaml" | "schema" | "model",
    readonly specPath: string,
    readonly issues: string[]
  ) {
    super(`[${phase.toUpperCase()}] ${issues.join("; ")}`);
    this.name = "SpecLoadError";
  }
}

// ── Loader ────────────────────────────────────────────────────────────────────

/**
 * Load and validate a .covenant.yaml through four sequential phases.
 *
 * Phase 1 -- FILE:   path exists and is readable
 * Phase 2 -- YAML:   valid YAML, top-level object
 * Phase 3 -- SCHEMA: JSON Schema Draft-07 validation (all errors collected)
 * Phase 4 -- MODEL:  Zod parse (skipped if Phase 3 has errors)
 *
 * @param specPath - Absolute or relative path to the .covenant.yaml file.
 * @returns Fully-validated CovenantSpec.
 * @throws SpecLoadError on any phase failure.
 */
export function loadSpec(specPath: string): CovenantSpec {
  const absPath = resolve(specPath);

  // Phase 1 -- FILE
  let text: string;
  try {
    text = readFileSync(absPath, "utf-8");
  } catch {
    throw new SpecLoadError("file", absPath, [`File not found: ${absPath}`]);
  }

  // Phase 2 -- YAML
  let raw: unknown;
  try {
    raw = yaml.load(text);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new SpecLoadError("yaml", absPath, [msg]);
  }

  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
    throw new SpecLoadError("yaml", absPath, [
      "YAML must be a mapping at the top level",
    ]);
  }

  // Phase 3 -- SCHEMA
  const validate = getValidator();
  const valid = validate(raw);
  if (!valid && validate.errors) {
    const issues = validate.errors.map(
      (e) => `${e.instancePath || "<root>"}: ${e.message ?? "invalid"}`
    );
    throw new SpecLoadError("schema", absPath, issues);
  }

  // Phase 4 -- MODEL
  try {
    return CovenantSpec.parse(raw);
  } catch (err) {
    if (err instanceof ZodError) {
      const issues = err.errors.map((e) => `${e.path.join(".")}: ${e.message}`);
      throw new SpecLoadError("model", absPath, issues);
    }
    throw err;
  }
}
